src/opt/Optimizer/Optimizer.hs

```diff
@@ -175,3 +180,90 @@ forwardGeneral = extend forwardGeneral'
                         ,pdefault= Nothing}
 
     forwardGeneral' e = e
+
+-- Desugars a for-loop into nested calls to map and flatMap and foreach:
+--
+-- for x <- listA, y <- listB, z <- ListC do
+--      fun
+-- end
+--
+-- into listA.flatMap(listB.flatMap(listC.map(fun)))
+--
+-- Credit: kaeluka for the use of foldl1 and zipWith in this manner
+forDesugared = extend forDesugared'
+  where
+    forDesugared' :: Expr -> Expr
+    forDesugared' e@For{emeta, sources, body} =
+      let
+        n = length sources
+        callNameList = if AST.AST.isCaptured e
+                      then replicate n (Name "foreach")
+                      else replicate (n-1) (Name "flatMap") ++ [Name "map"]
+        revSources = reverse sources
+        elemType = bodyType body
+        forprettyprint = nestCalls emeta callNameList sources body elemType
+      in
+        trace (show (ppExpr forprettyprint)) forprettyprint
+    forDesugared' e = e
+
+nestCalls :: Meta.Meta Expr -> [Name] -> [ForSource] -> Expr -> Type -> Expr -- nested MethodCalls and FunctionCalls
+nestCalls meta (name:_) (fs:[]) body elemType = intoCall meta name fs body elemType
+nestCalls meta (name:restOfNames) (fs:restFS) body elemType =
+  let nestedCall = intoCall meta name fs body elemType
+  in nestCalls meta restOfNames restFS nestedCall elemType
+
+intoCall :: Meta.Meta Expr -> Name -> ForSource -> Expr -> Type -> Expr -- MethodCall or FunctionCall
+intoCall met callName ForSource{forVar, forVarType, collection} bodyOrMethodCall elemType =
+  if isRefType (getType collection)
+  then let
+        param = [intoParam met Val forVar forVarType]
+        arguments = [intoClosure met param Nothing bodyOrMethodCall]
+       in
+        intoMethodCall met [elemType] collection callName arguments
+   else let
+        param = [intoParam met Val forVar forVarType]
+        arguments = [intoClosure met param Nothing bodyOrMethodCall] ++ [collection]
+       in
+        intoFunctionCall met [(fromMaybe intType forVarType), elemType] callName arguments
+
+-- helper functions
+bodyType body = getType body
+
+-- Maybe these should be in a kind of Util file, or in AST?
+intoClosure meta parameters mty body =
```

Comment:
I have mixed feelings about this functions. They would probably be useful for the Typechecker and others, but we are not using them there, we write the algebraic data type as it is. To maintain uniformity in the compiler, these ones should not be present. We can discuss this offline